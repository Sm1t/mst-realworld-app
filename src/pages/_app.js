import React from 'react';
import { Provider } from 'mobx-react';
import { getSnapshot } from 'mobx-state-tree';
import App, { Container } from 'next/app';

import { initializeStore } from '../store';
import { Layout } from '../layout';

export default class MyApp extends App {
  static async getInitialProps({ Component, ctx }) {
    //
    // Use getInitialProps as a step in the lifecycle when
    // we can initialize our store
    //
    const isServer = typeof window === 'undefined';
    const stores = initializeStore(isServer);

    //
    // Check whether the page being rendered by the App has a
    // static getInitialProps method and if so call it
    //
    let pageProps = {};

    if (Component.getInitialProps) {
      pageProps = await Component.getInitialProps(ctx, stores);
    }
    return {
      initialState: getSnapshot(stores),
      isServer,
      pageProps,
    };
  }

  constructor(props) {
    super(props);
    this.stores = initializeStore(props.isServer, props.initialState);
  }

  render() {
    const { Component, pageProps } = this.props;

    return (
      <Container>
        <Provider {...this.stores}>
          <Layout>
            <Component {...pageProps} />
          </Layout>
        </Provider>
      </Container>
    );
  }
}
