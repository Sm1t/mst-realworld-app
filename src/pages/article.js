import React, { Component } from 'react';
import { inject } from 'mobx-react';

import { ArticleService } from '../api/article.service';
import formatDate from '../utils/formatDate';

@inject(stores => ({ currentUser: stores.userStore.currentUser }))
class Article extends Component {
  static getInitialProps = async context => {
    const { slug } = context.query;
    const res = await ArticleService.get(slug);

    return { ...res.data.article };
  };

  render() {
    const {
      title,
      body,
      favoritesCount,
      createdAt,
      tagList,
      author: { username, image },
      currentUser,
    } = this.props;

    return (
      <div className="article-page">
        <div className="banner">
          <div className="container">
            <h1>{title}</h1>

            <div className="article-meta">
              <a href="/profile.html">
                <img src={image} alt="" />
              </a>
              <div className="info">
                <a href="/" className="author">
                  {username}
                </a>
                <span className="date">{formatDate(createdAt)}</span>
              </div>
              <button className="btn btn-sm btn-outline-secondary">
                <i className="ion-plus-round" />
                &nbsp; Follow {username}
              </button>
              &nbsp;&nbsp;
              <button className="btn btn-sm btn-outline-primary">
                <i className="ion-heart" />
                &nbsp; Favorite Post <span className="counter">{favoritesCount}</span>
              </button>
            </div>
          </div>
        </div>

        <div className="container page">
          <div className="row article-content">
            <div className="col-md-12">
              <p>{body}</p>
            </div>
          </div>

          <ul className="tag-list">
            {tagList.length > 0 && tagList.map((tag, index) => (
              <li key={index} className="tag-default tag-pill tag-outline">
                {tag}
              </li>
            ))}
          </ul>

          <hr />

          <div className="article-actions">
            <div className="article-meta">
              <a href="profile.html">
                <img src={image} alt="" />
              </a>
              <div className="info">
                <a href="/" className="author">
                  {username}
                </a>
                <span className="date">{formatDate(createdAt)}</span>
              </div>
              <button className="btn btn-sm btn-outline-secondary">
                <i className="ion-plus-round" />
                &nbsp; Follow {username}
              </button>
              &nbsp;
              <button className="btn btn-sm btn-outline-primary">
                <i className="ion-heart" />
                &nbsp; Favorite Post <span className="counter">{favoritesCount}</span>
              </button>
            </div>
          </div>

          <div className="row">
            <div className="col-xs-12 col-md-8 offset-md-2">
              {currentUser ? (
                <form className="card comment-form">
                  <div className="card-block">
                    <textarea
                      className="form-control"
                      placeholder="Write a comment..."
                      rows="3"
                    />
                  </div>
                  <div className="card-footer">
                    <img
                      src={currentUser.image}
                      className="comment-author-img"
                      alt=""
                    />
                    <button className="btn btn-sm btn-primary">
                      Post Comment
                    </button>
                  </div>
                </form>
              ) : (
                <p>
                  <a href="#/login">Sign in</a>&nbsp;or&nbsp;
                  <a href="#/register">sign up</a>
                  &nbsp;to add comments on this article.
                </p>
              )}

              {/* <div className="card">
                <div className="card-block">
                  <p className="card-text">
                    With supporting text below as a natural lead-in to additional
                    content.
                  </p>
                </div>
                <div className="card-footer">
                  <a href="" className="comment-author">
                    <img
                      src="http://i.imgur.com/Qr71crq.jpg"
                      className="comment-author-img"
                    />
                  </a>
                  &nbsp;
                  <a href="" className="comment-author">
                    Jacob Schmidt
                  </a>
                  <span className="date-posted">Dec 29th</span>
                </div>
              </div> */}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default Article;
