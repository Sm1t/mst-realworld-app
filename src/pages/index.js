import React from 'react';

import Articles from '../components/Articles';

const Index = () => (
  <div>
    <Articles />
  </div>
);

export default Index;