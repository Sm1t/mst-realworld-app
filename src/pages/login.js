import React, { useState } from 'react';
import { inject, observer } from 'mobx-react';

import Link from '../components/Link';
import ErrorsList from '../components/ErrorsList';

const ObservableErrorsList = observer(ErrorsList);

const selector = ({
  userStore: {
    errors, login,
  },
}) => ({
  errors,
  login,
});

export const Login = props => {
  const { errors, login } = props;
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const onSubmit = event => {
    event.preventDefault();
    login(email, password);
  };

  return (
    <div className="auth-page">
      <div className="container page">
        <div className="row">
          <div className="col-md-6 offset-md-3 col-xs-12">
            <h1 className="text-xs-center">Sign in</h1>
            <p className="text-xs-center">
              <Link href="/register">Need an account?</Link>
            </p>

            <ObservableErrorsList errors={errors} />

            <form onSubmit={onSubmit}>
              <fieldset className="form-group">
                <input
                  className="form-control form-control-lg"
                  type="email"
                  placeholder="Email"
                  value={email}
                  onChange={e => setEmail(e.target.value)}
                />
              </fieldset>
              <fieldset className="form-group">
                <input
                  className="form-control form-control-lg"
                  type="password"
                  placeholder="Password"
                  value={password}
                  onChange={e => setPassword(e.target.value)}
                />
              </fieldset>
              <button className="btn btn-lg btn-primary pull-xs-right">
                Sign up
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
};

export default inject(selector)(Login);
