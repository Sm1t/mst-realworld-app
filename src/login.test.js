import React from 'react';
import { shallow } from 'enzyme';

import { Login } from './pages/login';

describe('login', () => {
  it('should render without problems', () => {
    const wrapper = shallow(<Login />);

    expect(wrapper).toMatchSnapshot();
  });
  let wrapper;

  let loginFn;
  const fields = [
    { name: 'email', selector: 'input[type="email"]', value: 'mail@mail.ru' },
    { name: 'password', selector: 'input[type="password"]', value: 'pswrd123' },
  ];

  beforeEach(() => {
    loginFn = jest.fn(() => Promise.resolve());
    wrapper = shallow(<Login login={loginFn} />);
  });

  fields.forEach(({ name, selector, value }) => {
    it(`should properly update ${name} value`, () => {
      const field = wrapper.find(selector);

      expect(field.props().value).toEqual('');
      field.simulate('change', { target: { value } });
      expect(wrapper.find(selector).props().value).toEqual(
        value,
      );
    });
  });

  it('should call preventDefault on submit', () => {
    const preventDefault = jest.fn();

    wrapper.find('form').simulate('submit', { preventDefault });

    expect(preventDefault).toHaveBeenCalled();
  });

  it('should call login function with correct arguments', () => {
    const emailInput = wrapper.find(fields[0].selector);
    const passwordInput = wrapper.find(fields[1].selector);
    const email = fields[0].value;
    const password = fields[1].value;

    emailInput.simulate('change', { target: { value: email } });
    passwordInput.simulate('change', { target: { value: password } });

    wrapper.find('form').simulate('submit', { preventDefault() {} });

    expect(loginFn).toHaveBeenCalledWith(email, password);
  });
});
