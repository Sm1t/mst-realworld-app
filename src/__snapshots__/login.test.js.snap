// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login should render without problems 1`] = `
<div
  className="auth-page"
>
  <div
    className="container page"
  >
    <div
      className="row"
    >
      <div
        className="col-md-6 offset-md-3 col-xs-12"
      >
        <h1
          className="text-xs-center"
        >
          Sign in
        </h1>
        <p
          className="text-xs-center"
        >
          <withRouter(ActiveLink)
            href="/register"
          >
            Need an account?
          </withRouter(ActiveLink)>
        </p>
        <ErrorsList />
        <form
          onSubmit={[Function]}
        >
          <fieldset
            className="form-group"
          >
            <input
              className="form-control form-control-lg"
              onChange={[Function]}
              placeholder="Email"
              type="email"
              value=""
            />
          </fieldset>
          <fieldset
            className="form-group"
          >
            <input
              className="form-control form-control-lg"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
          </fieldset>
          <button
            className="btn btn-lg btn-primary pull-xs-right"
          >
            Sign up
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
